### PR TITLE
Fix s3 engine getting virtual columns

### LINKED ISSUE #35411: Virtual columns missing in s3 table function
With clickhouse/clickhouse-server:22.3.2.2 I can no longer access `_file` and `_path` virtual columns on s3() table function like I could with clickhouse/clickhouse-server:22.2.2.1. 

> A clear and concise description of what works not as it is supposed to.

This worked on previous release

```sql
SELECT _path FROM s3('https://host.domain.com/bucket/path/file.parquet') LIMIT 1;
-- Returns --
bucket/path/file.parquet
```

But instead with 22.3 I get

```sql
SELECT _path FROM s3('https://host.domain.com/bucket/path/file.parquet') LIMIT 1;
-- Returns --
Received exception from server (version 22.3.2):
Code: 8. DB::Exception: Received from localhost:9000. DB::Exception: Column '_path' is not presented in input data.: While executing ParquetBlockInputFormat: While executing S3. (THERE_IS_NO_COLUMN)
```

I also have the same behavior on `_file` virtual column. DESCRIBE and retireving data works as I would expect otherwise.

**How to reproduce**

* Which ClickHouse server version to use
    * 22.3.2.2 (`clickhouse/clickhouse-server:22.3.2.2` docker image) 
* Which interface to use, if matters
    * ` docker exec -it <clickhouse-container> clickhouse-client`
* Non-default settings, if any
    * Only added authentication to s3 endpoints
* `CREATE TABLE` statements for all tables involved
    * No tables created
* Sample data for all these tables, use [clickhouse-obfuscator](https://github.com/ClickHouse/ClickHouse/blob/master/programs/obfuscator/Obfuscator.cpp#L42-L80) if necessary
     * No cannot provide s3 link
* Queries to run that lead to unexpected result
    * `SELECT _path FROM s3('https://host.domain.com/bucket/path/file.parquet') LIMIT 1;`

**Expected behavior**

> A clear and concise description of what you expected to happen.

The same behavior as ClickHouse 22.2 that I get the path of the file that the row originates from with `_path` virtual column and the file name with `_file`.

**Error message and/or stacktrace**

```
2022.03.18 15:06:09.951316 [ 45 ] {e874adec-e4ec-4d97-95aa-cc63273472fa} <Error> TCPHandler: Code: 8. DB::Exception: Column '_path' is not presented in input data.: While executing ParquetBlockInputFormat: While executing S3. (THERE_IS_NO_COLUMN), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xa4dde1a in /usr/bin/clickhouse
1. DB::Exception::Exception<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>(int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0xa54a504 in /usr/bin/clickhouse
2. DB::ArrowColumnToCHColumn::getMissingColumns(arrow::Schema const&) const @ 0x154d358f in /usr/bin/clickhouse
3. DB::ParquetBlockInputFormat::prepareReader() @ 0x1558744e in /usr/bin/clickhouse
4. DB::ParquetBlockInputFormat::generate() @ 0x15586c72 in /usr/bin/clickhouse
5. DB::ISource::tryGenerate() @ 0x1548cdf5 in /usr/bin/clickhouse
6. DB::ISource::work() @ 0x1548c9ba in /usr/bin/clickhouse
7. DB::ExecutionThreadContext::executeTask() @ 0x154ad143 in /usr/bin/clickhouse
8. DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) @ 0x154a0b9e in /usr/bin/clickhouse
9. DB::PipelineExecutor::executeStep(std::__1::atomic<bool>*) @ 0x154a03c0 in /usr/bin/clickhouse
10. DB::PullingPipelineExecutor::pull(DB::Chunk&) @ 0x154b1c6e in /usr/bin/clickhouse
11. DB::StorageS3Source::generate() @ 0x14c8774c in /usr/bin/clickhouse
12. DB::ISource::tryGenerate() @ 0x1548cdf5 in /usr/bin/clickhouse
13. DB::ISource::work() @ 0x1548c9ba in /usr/bin/clickhouse
14. DB::SourceWithProgress::work() @ 0x156e4282 in /usr/bin/clickhouse
15. DB::ExecutionThreadContext::executeTask() @ 0x154ad143 in /usr/bin/clickhouse
16. DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) @ 0x154a0b9e in /usr/bin/clickhouse
17. ? @ 0x154a2504 in /usr/bin/clickhouse
18. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xa584c97 in /usr/bin/clickhouse
19. ? @ 0xa58881d in /usr/bin/clickhouse
20. ? @ 0x7f4462478609 in ?
21. __clone @ 0x7f446239d163 in ?
```

**Additional context**

> Add any other context about the problem here.

I did not test this with Amazon S3 in this case but instead with a local deployment of Minio running docker image `minio/minio:RELEASE.2020-02-07T04-56-50Z`. This is an old version but it worked fine with the old ClickHouse release. I will try with a real Amazon S3 endpoint as soon as I am able to.

### PR DESCRIPTION
**Pull request by kssenii**
Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix s3 engine getting virtual columns. Closes https://github.com/ClickHouse/ClickHouse/issues/35411.

### CODE REVIEW

File: src/Storages/StorageS3.cpp [hunk 1]
```diff
@@ -660,8 +660,14 @@ Pipe StorageS3::read(
     Block block_for_format;
     if (isColumnOriented())
     {
+        auto fetch_columns = column_names;
+        fetch_columns.erase(std::remove_if(fetch_columns.begin(), fetch_columns.end(),
+                              [](const String & col){return col == "_path" || col == "_file"; }));
```
Comment (alesapin) [inline src/Storages/StorageS3.cpp:665]: Maybe it's better to define list of virtual columns somewhere?

Comment (kssenii) [reply src/Storages/StorageS3.cpp]: done

File: src/Storages/StorageS3.cpp [hunk 2]
```diff
@@ -660,8 +660,19 @@ Pipe StorageS3::read(
     Block block_for_format;
     if (isColumnOriented())
     {
+        auto fetch_columns = column_names;
+        const auto & virtuals = getVirtuals();
+        fetch_columns.erase(
+            std::remove_if(
+                fetch_columns.begin(), fetch_columns.end(),
+                [&](const String & col)
+                { return std::any_of(virtuals.begin(), virtuals.end(), [&](const NameAndTypePair & virtual_col){ return col == virtual_col.name; }); }));
```
Comment (antonio2368) [inline src/Storages/StorageS3.cpp:669]: Just a good thing to know, there is [std::erase_if](https://en.cppreference.com/w/cpp/container/vector/erase2)

File: src/Storages/StorageS3.cpp [hunk 3]
```diff
@@ -660,8 +660,18 @@ Pipe StorageS3::read(
     Block block_for_format;
     if (isColumnOriented())
     {
+        auto fetch_columns = column_names;
+        const auto & virtuals = getVirtuals();
```
Comment (alexey-milovidov) [inline src/Storages/StorageS3.cpp:664]: Does it mean that virtual columns shadow ordinary columns?
But it's better to have opposite: if a file has `_path` column, then this column will be not virtual, and the virtual `_path` column will be unaccessible.

Comment (kssenii) [reply src/Storages/StorageS3.cpp:664]: Done

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**